dlt/destinations/impl/clickhouse/sql_client.py

```diff
@@ -0,0 +1,210 @@
+            )
+
+    def _list_tables(self) -> List[str]:
+        rows = self.execute_sql(
+            """
+            SELECT name
+            FROM system.tables
+            WHERE database = %s
+            AND name LIKE %s
+            """,
+            (
+                self.database_name,
+                f"{self.dataset_name}{self.credentials.dataset_table_separator}%",
+            ),
+        )
+        return [row[0] for row in rows]
+
+    @contextmanager
+    @raise_database_error
+    def execute_query(
+        self, query: AnyStr, *args: Any, **kwargs: Any
+    ) -> Iterator[ClickHouseDBApiCursorImpl]:
+        assert isinstance(query, str), "Query must be a string."
+
+        db_args = kwargs.copy()
+
+        if args:
+            query, db_args = _convert_to_old_pyformat(query, args, OperationalError)
+            db_args.update(kwargs)
+
+        with self._conn.cursor() as cursor:
+            for query_line in query.split(";"):
+                if query_line := query_line.strip():
+                    try:
+                        cursor.execute(query_line, db_args)
+                    except KeyError as e:
+                        raise DatabaseTransientException(OperationalError()) from e
+
+            yield ClickHouseDBApiCursorImpl(cursor)  # type: ignore[abstract]
+
+    def fully_qualified_dataset_name(self, escape: bool = True) -> str:
+        database_name = self.database_name
+        dataset_name = self.dataset_name
+        if escape:
+            database_name = self.capabilities.escape_identifier(database_name)
+            dataset_name = self.capabilities.escape_identifier(dataset_name)
+        return f"{database_name}.{dataset_name}"
+
+    def make_qualified_table_name(self, table_name: str, escape: bool = True) -> str:
+        database_name = self.database_name
+        table_name = f"{self.dataset_name}{self.credentials.dataset_table_separator}{table_name}"
+        if escape:
+            database_name = self.capabilities.escape_identifier(database_name)
+            table_name = self.capabilities.escape_identifier(table_name)
+        return f"{database_name}.{table_name}"
+
+    @classmethod
+    def _make_database_exception(cls, ex: Exception) -> Exception:
+        if isinstance(ex, clickhouse_driver.dbapi.errors.OperationalError):
+            if "Code: 57." in str(ex) or "Code: 82." in str(ex) or "Code: 47." in str(ex):
+                return DatabaseTerminalException(ex)
+            elif "Code: 60." in str(ex) or "Code: 81." in str(ex):
+                return DatabaseUndefinedRelation(ex)
+            else:
+                return DatabaseTransientException(ex)
+        elif isinstance(
+            ex,
+            (
+                clickhouse_driver.dbapi.errors.OperationalError,
+                clickhouse_driver.dbapi.errors.InternalError,
+            ),
+        ):
+            if term := cls._maybe_make_terminal_exception_from_data_error():
```

Comment:
this is in a couple of destinations but does not do anything for most. I'll keep it for now, but some destination overall cleanup is due sometime i think.

Reply:
ok, I kicked it out, let's see if all the tests still pass